### ISSUE TITLE
Critical problem with non-squared image

### ISSUE THREAD

**Issue by rroseand** [allanzelener/YAD2K#86]
Hello,

First of all thanks for this implementation. 
I found a problem with the code when training non-square images. 
The problem came from the "yolo_head" function :
conv_dims is equal to [height, width] and is used to divide box_xy where the order is reversed (x is along the width axis and y along the height).

This causes the network to converge towards boxes which doesn't make sense at all.
The problem is solved by simply reversing conv_dims before dividing box_xy and box_wh with tf.reverse

Thanks again and i hope it helps

**Comment by joeysu** [comment]
Thanks @rroseand 

**conv_height_index = K.arange(0, stop=conv_dims[0])
conv_width_index = K.arange(0, stop=conv_dims[1])**

to

**conv_width_index = K.arange(0, stop=conv_dims[0])
conv_height_index = K.arange(0, stop=conv_dims[1])**

Is that right?

**Comment by rroseand** [comment]
Hello, I don't know if it will work that way because the conv_index is also creating using those. 
I didn't quite understand how the conv_index was created so I did it again myself during my search for the error.

Maybe it will work by just putting :
**conv_dims = tf.reverse(conv_dims, [-1])**
just before the line
**conv_dims = K.cast(K.reshape(conv_dims, [1, 1, 1, 1, 2]), K.dtype(feats))**

or you can also copy my code that changed the way that conv_index is created :

    conv_height_index = K.arange(0, stop=conv_dims[0])
    conv_width_index = K.arange(0, stop=conv_dims[1])
    conv_height_index = tf.transpose(tf.tile(tf.expand_dims(conv_height_index, 0), [conv_dims[1], 1]))

    conv_width_index = tf.tile(tf.expand_dims(conv_width_index, 0), [conv_dims[0], 1])
    conv_index = tf.stack([conv_width_index, conv_height_index], -1)
    conv_index = tf.reshape(conv_index, [1, conv_dims[0], conv_dims[1], 1, 2])
    conv_index = K.cast(conv_index, K.dtype(feats))

    feats = K.reshape(
        feats, [-1, conv_dims[0], conv_dims[1], num_anchors, num_classes + 5])

    conv_dims = tf.reverse(conv_dims, [-1])
    conv_dims = K.cast(K.reshape(conv_dims, [1, 1, 1, 1, 2]), K.dtype(feats))

I did it in tensorflow though, i think it can be done in pure keras.

**Comment by joeysu** [comment]
@rroseand 
Hi, sorry for the late replay.

Your reverse code works.
Thanks again!

The way I said before predicted some weird bboxes when I trained with one signal image to overfit.

**Comment by tianyu-tristan** [comment]
@s5plus1 Hi, sorry to bother you as I'm not sure if you're still following this track. Have you successfully tested it with non-square image by overfitting a small set? If you did, is this the only issue you had fixed?
